File: Week_3/Python/bfs.py
from queue import Queue

#  Graph representation using adjacency list
n = 5 # no of vertices
e = 7 # no of edges
# using 1-based indexing thus the vertices are 1,2,3,...n

adj = [[] for i in range(n+1)]
edges = [[1,2],[2,3],[1,4],[4,5],[3,4],[3,5],[2,5]]

for x in edges:
    adj[x[0]].append(x[1]); # adding vertex x[1] to adjacent list of vertex x[0]
    adj[x[1]].append(x[0]); # adding vertex x[0] to adjacent list of vertex x[1]

vis = [False for i in range(n+1)]
q = Queue() # A queue to store the order of vertex visit
q.put(2)
vis[2] = True

# BFS on a graph

while (q.qsize() != 0):
    vertex = q.get()
    print(vertex,end = ' ')
    for child in adj[vertex]:
        if not vis[child]:
            q.put(child)
            vis[child] = True
print()

File: Week_3/Cpp/bfs.cpp
#include<bits/stdc++.h>
using namespace std;

typedef long long ll;

int main() {
    // Graph representation using adjacency list
    ll n = 5; // no of vertices
    ll e = 7; // no of edges
    // using 1-based indexing thus the vertices are 1,2,3,...n
    vector<vector<ll>> adj(n+1); // declaring adjacency list

    vector<pair<ll,ll>> edges = {{1,2},{2,3},{1,4},{4,5},{3,4},{3,5},{2,5}};

    for(auto x:edges){
        ll a = x.first;
        ll b = x.second;
        adj[a].push_back(b); // adding vertex b to adjacent list of vertex a
        adj[b].push_back(a); // adding vertex a to adjacent list of vertex b
    }
    
    // vis will store if the vertex i is visited for index i
    vector<bool> vis(n+1,false);

    // this will print the sequence of vertices which our bfs functions visits
    queue<ll> q;
    q.push(1);
    vis[1] = true;
    while(!q.empty()){
        ll vertex = q.front();
        cout << vertex << " ";
        q.pop();
        for(auto child:adj[vertex]){
            if(!vis[child]){
                q.push(child);
                vis[child] = true;
            }
        }
    }  
}